midend/interpreter.cpp

```diff
@@ -913,6 +932,220 @@ void ExpressionEvaluator::postorder(const IR::BoolLiteral* expression) {
     set(expression, new SymbolicBool(expression));
 }
 
+void ExpressionEvaluator::postorder(const IR::Equ* expression) {
+    auto l = get(expression->left);
+    auto clone = expression->clone();
+    DoConstantFolding cf(refMap, typeMap);
+    cf.setCalledBy(this);
+    if (l->is<SymbolicInteger>()) {
+        auto lValue = clone->left->to<IR::Constant>();
+        auto rValue = clone->right->to<IR::Constant>();
+        if (!lValue) {
+            clone->left = get(clone->left)->to<SymbolicInteger>()->constant;
+            lValue = clone->left->to<IR::Constant>();
+        }
+        if (!rValue) {
+            clone->right = get(clone->right)->to<SymbolicInteger>()->constant;
+            rValue = clone->right->to<IR::Constant>();
+        }
+        BUG_CHECK(clone->left, "%1%: expected a constant", clone->left);
+        BUG_CHECK(clone->right, "%1%: expected a constant", clone->right);
+        if (lValue->value == rValue->value) {
+            set(expression, new SymbolicBool(new IR::BoolLiteral(1)));
+            return;
+        } else {
+            set(expression, new SymbolicBool(new IR::BoolLiteral(0)));
+            return;
+        }
+    }else if (l->is<SymbolicBool>()) {
+        auto lValue = clone->left->to<IR::BoolLiteral>();
+        auto rValue = clone->right->to<IR::BoolLiteral>();
+        if (!lValue) {
+            clone->left = new IR::BoolLiteral(get(clone->left)->to<SymbolicBool>()->value);
+            lValue = clone->left->to<IR::BoolLiteral>();
+        }
+        if (!rValue) {
+            clone->right = new IR::BoolLiteral(get(clone->right)->to<SymbolicBool>()->value);
+            rValue = clone->right->to<IR::BoolLiteral>();
+        }
+        BUG_CHECK(clone->left, "%1%: expected a boolean", clone->left);
+        BUG_CHECK(clone->right, "%1%: expected a boolean", clone->right);
+        if (lValue->value == rValue->value) {
+            set(expression, new SymbolicBool(new IR::BoolLiteral(1)));
+            return;
+        } else {
+            set(expression, new SymbolicBool(new IR::BoolLiteral(0)));
+            return;
+        }
+    }
+    BUG("%1%: unexpected type", l);
+}
+
+void ExpressionEvaluator::postorder(const IR::Neq* expression) {
+    auto l = get(expression->left);
+    auto clone = expression->clone();
+    DoConstantFolding cf(refMap, typeMap);
+    cf.setCalledBy(this);
```

Comment:
This makes no sense. You do not call the constant folding pass. Instead of do constant folding manually.
There is no reason to have such a code duplication. 
This is true for all the cases below.